### PR TITLE
mitigate rounding errors

### PR DESCRIPTION
**Pull request by trmid**
- implement dust collector strategy
- add yield buffer
- also fixes GEN-1073

### CODE REVIEW

File: src/PrizeVaultFactory.sol [hunk 1]
```diff
@@ -54,17 +54,19 @@ contract PrizeVaultFactory {
...
      * @param _yieldFeePercentage Yield fee percentage
+     * @param _yieldBuffer Amount of yield to keep as a buffer
      * @param _owner Address that will gain ownership of this contract
      * @return PrizeVault The newly deployed PrizeVault
      */
     function deployVault(
-      string calldata _name,
-      string calldata _symbol,
+      string memory _name,
+      string memory _symbol,
```
Comment (trmid) [inline src/PrizeVaultFactory.sol:63]: Switched this to `memory` to avoid `via-ir` compilation.

File: test/contracts/mock/YieldVault.sol [hunk 2]
```diff
@@ -16,19 +16,19 @@ contract YieldVault is ERC4626Mock {
...
-        uint256 supply = totalSupply();
-        return (shares == 0 || supply == 0) ? shares : shares.mulDiv(totalAssets(), supply, rounding);
-    }
+    // function _convertToAssets(
+    //     uint256 shares,
+    //     Math.Rounding rounding
+    // ) internal view virtual override returns (uint256) {
+    //     uint256 supply = totalSupply();
+    //     return (shares == 0 || supply == 0) ? shares : shares.mulDiv(totalAssets(), supply, rounding);
+    // }
```
Comment (trmid) [inline test/contracts/mock/YieldVault.sol:33]: The prize vault is now able to handle small differences in conversions

File: src/PrizeVault.sol [hunk 3]
```diff
@@ -284,48 +300,57 @@ contract PrizeVault is TwabERC20, Claimable, IERC4626, ILiquidationSource, Ownab
...
         }
     }
 
     /// @inheritdoc IERC4626
     /// @dev Considers the uint96 limit on total share supply in the TwabController
+    /// @dev TODO: add reasoning for exclusion of latent balance
+    /// @dev TODO: return 0 if deposits and mints are disabled due to less assets than supply
     function maxDeposit(address) public view returns (uint256) {
         // the vault will never mint more than 1 share per asset, so no need to convert supply buffer to assets
         uint256 _supplyBuffer = type(uint96).max - totalSupply();
```
Comment (trmid) [inline src/PrizeVault.sol]: consider other naming ideas like "supplyAvailable"?

File: src/PrizeVault.sol [hunk 4]
```diff
@@ -284,48 +300,57 @@ contract PrizeVault is TwabERC20, Claimable, IERC4626, ILiquidationSource, Ownab
...
-        return _ownerAssets < _maxAssetWithdraw ? _ownerAssets : _maxAssetWithdraw;
+        return _ownerAssets < _maxWithdraw ? _ownerAssets : _maxWithdraw;
     }
 
     /// @inheritdoc IERC4626
+    /// @dev TODO: add reasoning for inclusion of latent balance
     function maxRedeem(address _owner) public view returns (uint256) {
         // the owner will never receive more than 1 asset per share, so no need to convert max withdraw to shares
-        uint256 _maxShareRedemption = yieldVault.maxWithdraw(address(this));
+        uint256 _maxRedeem = yieldVault.maxWithdraw(address(this)) + _asset.balanceOf(address(this));
```
Comment (trmid) [inline src/PrizeVault.sol:351]: Should convert max withdraw assets to shares rounding up to ensure they get the most out of the limits.

File: src/PrizeVault.sol (hunk 3 above)
Comment (trmid) [reply src/PrizeVault.sol]: changed here: [62d6566](https://github.com/GenerationSoftware/pt-v5-vault/pull/69/commits/62d656671d780bd445df2de1328dfa2ea72fe8ba)

File: src/PrizeVault.sol [hunk 5]
```diff
@@ -284,48 +300,57 @@ contract PrizeVault is TwabERC20, Claimable, IERC4626, ILiquidationSource, Ownab
...
+        uint256 _maxWithdraw = yieldVault.maxWithdraw(address(this)) + _asset.balanceOf(address(this));
 
         // the owner may receive less than 1 asset per share, so we must convert their balance here
         uint256 _ownerAssets = convertToAssets(balanceOf(_owner));
-        return _ownerAssets < _maxAssetWithdraw ? _ownerAssets : _maxAssetWithdraw;
+        return _ownerAssets < _maxWithdraw ? _ownerAssets : _maxWithdraw;
     }
 
     /// @inheritdoc IERC4626
+    /// @dev TODO: add reasoning for inclusion of latent balance
```
Comment (asselstine) [inline src/PrizeVault.sol:348]: Comment looks wrong

File: src/PrizeVault.sol [hunk 6]
```diff
@@ -284,48 +300,57 @@ contract PrizeVault is TwabERC20, Claimable, IERC4626, ILiquidationSource, Ownab
...
     function maxRedeem(address _owner) public view returns (uint256) {
         // the owner will never receive more than 1 asset per share, so no need to convert max withdraw to shares
-        uint256 _maxShareRedemption = yieldVault.maxWithdraw(address(this));
+        uint256 _maxRedeem = yieldVault.maxWithdraw(address(this)) + _asset.balanceOf(address(this));
```
Comment (asselstine) [inline src/PrizeVault.sol:351]: should be `maxWithdraw`, not maxRedeem eh?

File: src/PrizeVault.sol [hunk 7]
```diff
@@ -284,48 +300,57 @@ contract PrizeVault is TwabERC20, Claimable, IERC4626, ILiquidationSource, Ownab
...
         uint256 _ownerShares = balanceOf(_owner);
-        return _ownerShares < _maxShareRedemption ? _ownerShares : _maxShareRedemption;
+        return _ownerShares < _maxRedeem ? _ownerShares : _maxRedeem;
```
Comment (asselstine) [inline src/PrizeVault.sol:353]: Is wildly inaccurate when undercollateralized. For example:

If vault share totalSupply = 100 and yield vault asset balance = 50

- The first call will return 50, even though the user can burn 100 of their vault shares. Let's say they burn 50 of their shares now. This leaves 25 yield vault asset balance, and 50 vault total supply.
- The second call will return 25, which is only half what the user can redeem.

I think this is actually related to your comment above ^^

**Pull request closed without merging**